core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.ts

```diff
@@ -170,6 +182,138 @@ export class ContributionAndReviewService {
+              suggestion: suggestion,
+              details: (
+                fetchSuggestionsResponse.target_id_to_opportunity_dict[
+                  suggestion.target_id])
+            };
+            responseSuggestionIdToDetails[
+              suggestion.suggestion_id] = suggestionDetails;
+          });
+          return {
+            suggestionIdToDetails: responseSuggestionIdToDetails,
+            more: false
+          };
+        })
+    );
+  }
+
+  sortTranslationSuggestionsByState(
+      translationSuggestions: SuggestionBackendDict[],
+      states: States,
+      initStateName: string | null
+  ): SuggestionBackendDict[] {
+    if (!initStateName) {
+      return translationSuggestions;
+    }
+
+    const stateNamesInOrder = this.computeGraphService.
+      computeBfsTraversalOfStates(
+        initStateName,
+        states,
+        initStateName
+      );
+    const translationSuggestionsByState = (
+      ContributionAndReviewService
+        .groupTranslationSuggestionsByState(translationSuggestions));
+    const sortedTranslationCards: SuggestionBackendDict[] = [];
+
+    for (const stateName of stateNamesInOrder) {
+      const cardsForState = (
+        translationSuggestionsByState.get(stateName) || []);
+      cardsForState.sort(
+        ContributionAndReviewService.compareTranslationSuggestions);
+      sortedTranslationCards.push(...cardsForState);
+    }
+    return sortedTranslationCards;
+  }
+
+  static groupTranslationSuggestionsByState(
+      translationSuggestions: SuggestionBackendDict[])
+    : Map<string, SuggestionBackendDict[]> {
+    const translationSuggestionsByState = new Map<
+    string, SuggestionBackendDict[]>();
+
+    for (const translationSuggestion of translationSuggestions) {
+      const stateName = translationSuggestion.change.state_name;
+      const suggestionsForState = translationSuggestionsByState.get(
+        stateName) || [];
+      suggestionsForState.push(translationSuggestion);
+      translationSuggestionsByState.set(stateName, suggestionsForState);
+    }
+    return translationSuggestionsByState;
+  }
+
+  // Compares translation suggestions based on type and index.
+  static compareTranslationSuggestions(
+      cardA: SuggestionBackendDict,
+      cardB: SuggestionBackendDict
+  ): number {
+    const cardATypeOrder = ContributionAndReviewService.
+      getTypeOrder(cardA.change.content_id);
+    const cardBTypeOrder = ContributionAndReviewService
+      .getTypeOrder(cardB.change.content_id);
+
+    if (cardATypeOrder !== cardBTypeOrder) {
+      return cardATypeOrder - cardBTypeOrder;
+    } else {
+      const cardAIndex = ContributionAndReviewService
+        .getIndex(cardA.change.content_id);
+      const cardBIndex = ContributionAndReviewService
+        .getIndex(cardB.change.content_id);
+
+      return cardAIndex - cardBIndex;
+    }
+  }
+
+  // Returns the type order for a given content ID.
+  static getTypeOrder(contentId: string): number {
```

Comment:
```suggestion
  static getTranslationContentTypeOrder(contentId: string): number {
```
Since this method is declared at the class level, I would include a bit more context in the name. Ditto below.

Reply:
Done